src/pkg/backup/details/testdata/in_deets.go

```diff
@@ -0,0 +1,372 @@
+	// so long as they are consistent for the test in question
+	Locations map[string]map[string]struct{}
+	Deleted   map[string]map[string]struct{}
+}
+
+func newLocSet() *locSet {
+	return &locSet{
+		Locations: map[string]map[string]struct{}{},
+		Deleted:   map[string]map[string]struct{}{},
+	}
+}
+
+func (ls *locSet) AddItem(locationRef, itemRef string) {
+	if _, ok := ls.Locations[locationRef]; !ok {
+		ls.Locations[locationRef] = map[string]struct{}{}
+	}
+
+	ls.Locations[locationRef][itemRef] = exists
+	delete(ls.Deleted[locationRef], itemRef)
+}
+
+func (ls *locSet) RemoveItem(locationRef, itemRef string) {
+	delete(ls.Locations[locationRef], itemRef)
+
+	if _, ok := ls.Deleted[locationRef]; !ok {
+		ls.Deleted[locationRef] = map[string]struct{}{}
+	}
+
+	ls.Deleted[locationRef][itemRef] = exists
+}
+
+func (ls *locSet) AddLocation(locationRef string) {
+	if _, ok := ls.Locations[locationRef]; !ok {
+		ls.Locations[locationRef] = map[string]struct{}{}
+	}
+	// don't purge previously deleted items, or child locations.
+	// Assumption is that their itemRef is unique, and still deleted.
+	delete(ls.Deleted, locationRef)
+}
+
+func (ls *locSet) RemoveLocation(locationRef string) {
+	ss := ls.Subset(locationRef)
+
+	for lr := range ss.Locations {
+		items := ls.Locations[lr]
+
+		delete(ls.Locations, lr)
+
+		if _, ok := ls.Deleted[lr]; !ok {
+			ls.Deleted[lr] = map[string]struct{}{}
+		}
+
+		for ir := range items {
+			ls.Deleted[lr][ir] = exists
+		}
+	}
+}
+
+// MoveLocation takes the LAST elemet in the fromLocation (and all)
+// children matching the prefix, and relocates it as a child of toLocation.
+// ex: MoveLocation("/a/b/c", "/d") will move all entries with the prefix
+// "/a/b/c" into "/d/c".  This also deletes all "/a/b/c" entries and children.
+// assumes item IDs don't change across the migration.  If item IDs do change,
+// that difference will need to be handled manually by the caller.
+// returns the base folder's new location (ex: /d/c)
+func (ls *locSet) MoveLocation(fromLocation, toLocation string) string {
+	ss := ls.Subset(fromLocation)
+	lastElem := path.Builder{}.Append(path.Split(fromLocation)...).LastElem()
+
+	for lr, items := range ss.Locations {
+		newLoc := toLocation + "/" + lastElem
+		trimLoc := strings.TrimPrefix(lr, fromLocation)
+		trimLoc = strings.TrimPrefix(trimLoc, "/")
+
+		if len(trimLoc) > 0 {
+			newLoc += "/" + trimLoc
+		}
+
+		for ir := range items {
+			ls.RemoveItem(lr, ir)
+			ls.AddItem(newLoc, ir)
+		}
+
+		ls.RemoveLocation(lr)
+		ls.AddLocation(newLoc)
+	}
+
+	if len(toLocation) == 0 {
+		return lastElem
+	}
+
+	return toLocation + "/" + lastElem
+}
+
+func (ls *locSet) RenameLocation(fromLocation, toLocation string) {
+	ss := ls.Subset(fromLocation)
+
+	for lr, items := range ss.Locations {
+		newLoc := toLocation
+		trimLoc := strings.TrimPrefix(lr, fromLocation)
+		trimLoc = strings.TrimPrefix(trimLoc, "/")
+
+		if len(trimLoc) > 0 {
+			newLoc += "/" + trimLoc
+		}
```

Comment:
I feel like a lot of this could be simplified by converting to a path.Builder, calling UpdateParent, and converting back to a string

Reply:
ooh, fancy.